From: YOSHIFUJI Hideaki <yoshfuji@linux-ipv6.org>
Date: Wed, 14 Nov 2012 17:37:46 +0000 (+0900)
Subject: ping,ping6: Do not assume radix point is denoted by '.' (-i option).
X-Git-Url: http://www.linux-ipv6.org/gitweb/gitweb.cgi?p=gitroot%2Fiputils.git;a=commitdiff_plain;h=04d3f79834957dc760850b00b8b52efa14e74757

ping,ping6: Do not assume radix point is denoted by '.' (-i option).

In some loacles (e.g., en, ja), radix point is denoted by '.'.
In other locales (e.g., de, fr), radix point is denoted by ','.

The argument parser for interval (-i option) checked '.' to
determine if the argument is integer or not, and used sscanf
with %d or %f.  As we are enabling locale support along with IDN
support, such an assumption is not good at all.

Just let strtod(3) handle difference of locales.

Signed-off-by: YOSHIFUJI Hideaki <yoshfuji@linux-ipv6.org>
---

diff --git a/ping_common.c b/ping_common.c
index af8cabc..56cc967 100644
--- a/ping_common.c
+++ b/ping_common.c
@@ -1,6 +1,7 @@
 #include "ping_common.h"
 #include <ctype.h>
 #include <sched.h>
+#include <math.h>
 
 int options;
 
@@ -255,24 +256,20 @@ void common_options(int ch)
 		break;
 	case 'i':		/* wait between sending packets */
 	{
-		if (strchr(optarg, '.')) {
-			float t;
-			if (sscanf(optarg, "%f", &t) != 1) {
-				fprintf(stderr, "ping: bad timing interval.\n");
-				exit(2);
-			}
-			interval = (int)(t*1000);
-		} else if (sscanf(optarg, "%d", &interval) == 1) {
-			interval *= 1000;
-		} else {
-			fprintf(stderr, "ping: bad timing interval.\n");
-			exit(2);
-		}
+		double dbl;
+		char *ep;
 
-		if (interval < 0) {
-			fprintf(stderr, "ping: bad timing interval.\n");
+		errno = 0;
+		dbl = strtod(optarg, &ep);
+
+		if (errno || *ep != '\0' ||
+		    !finite(dbl) || dbl < 0.0 || dbl >= (double)INT_MAX / 1000 - 1.0) {
+			fprintf(stderr, "ping: bad timing interval\n");
 			exit(2);
 		}
+
+		interval = (int)(dbl * 1000);
+
 		options |= F_INTERVAL;
 		break;
 	}
